Kustomize docs centralization (#33178)
sourcegraph/sourcegraph: doc/admin/install/kubernetes/kustomize.md

- Better enable Sourcegraph to support you if you run into issues, because how your deployment varies from our defaults is encapsulated in a small set of files.

## Using Kustomize

### General premise

In general, we recommend that customizations work like this:

1. Create, customize, and apply overlays for your deployment
2. Ensure the services came up correctly, then commit all the customizations to the new branch

  ```sh
  git add /overlays/$MY_OVERLAY/*
  # Keeping all overlays contained to a single commit allows for easier cherry-picking
  git commit amend -m "overlays: update $MY_OVERLAY"
  ```

See the [overlays guide](#overlays) to learn about the [overlays we provide](#provided-overlays) and [how to create your own overlays](#custom-overlays).

## Overlays

An [*overlay*](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/#bases-and-overlays) specifies customizations for a base directory of Kubernetes manifests, in this case the `base/` directory in the [reference repository](#reference-repository). 

Overlays can:

> - Be used for example to change the number of replicas, change a namespace, add a label, etc

Author: ```suggestion

- Be used for example to change the number of replicas, change a namespace, add a label, etc
```

- Refer to other overlays that eventually refer to the base (forming a directed acyclic graph with the base as the root)

### Using overlays

Overlays can be used in one of two ways:

> - With `kubectl`: Starting with `kubectl` client version 1.14 `kubectl` can handle `kustomization.yaml` files directly.

Author: ```suggestion

- With `kubectl`: Starting with `kubectl` client version 1.14 `kubectl` can handle `kustomization.yaml` files directly.
```

command is a directory containing a `kustomization.yaml` file.
- With`kustomize`: This generates manifest files that are then applied in the conventional way using `kubectl apply -f`.

The overlays provided in our [overlays directory](https://github.com/sourcegraph/deploy-sourcegraph/tree/master/overlays) rely on the `kustomize` tool and the `overlay-generate-cluster.sh` script in the
`root` directory to generate the manifests. There are two reasons why it was set up like this:

- It avoids having to put a `kustomization.yaml` file in the `base` directory and forcing users that don't use overlays
  to deal with it (unfortunately `kubectl apply -f` doesn't work if a `kustomization.yaml` file is in the directory).
- It generates manifests instead of applying them directly. This provides opportunity to additionally validate the files
  and also allows using `kubectl apply -f` with `--prune` flag turned on (`apply -k` with `--prune` does not work correctly).

### Generating Manifests

To generate Kubernetes manifests from an overlay, run the `overlay-generate-cluster.sh` with two arguments:

> - the name of the overlay

Author: ```suggestion

- the name of the overlay
```

- and a path to an output directory where the generated manifests will be

For example:
```sh
#                overlay directory name    output directory
#                                 |             |
./overlay-generate-cluster.sh my-overlay generated-cluster
```

After executing the script you can apply the generated manifests from the `generated-cluster` directory:

```sh
kubectl apply --prune -l deploy=sourcegraph -f generated-cluster --recursive
```

We recommend that you:

> - [Update the `./overlay-generate-cluster` script](./operations.md#applying-manifests) to apply the generated manifests from the `generated-cluster` directory with something like the above snippet

Author: ```suggestion

- [Update the `./overlay-generate-cluster` script](./operations.md#applying-manifests) to apply the generated manifests from the `generated-cluster` directory with something like the above snippet
```

# Kustomize

## About Kustomize

Sourcegraph supports the use of [Kustomize](https://kustomize.io) to modify and customize our Kubernetes manifests. Kustomize is a template-free way to customize configuration with a simple configuration file.

Some benefits of using Kustomize to generate manifests instead of modifying the base directly include:

- Reduce the odds of encountering a merge conflict when [updating Sourcegraph](update.md) - they allow you to separate your unique changes from the upstream base files Sourcegraph provides.

> - Better enable Sourcegraph to support you if you run into issues, because how your deployment varies from our [reference deployment](#reference-repository) is encapsulated in a small set of files.

Author: ```suggestion
- Better enable Sourcegraph to support you if you run into issues, because how your deployment varies from our defaults is encapsulated in a small set of files.
```

## Using Kustomize

### General premise

In general, we recommend that customizations work like this:

1. Create, customize, and apply overlays for your deployment
2. Ensure the services came up correctly, then commit all the customizations to the new branch

  ```sh
  git add /overlays/$MY_OVERLAY/*
  # Keeping all overlays contained to a single commit allows for easier cherry-picking
  git commit amend -m "overlays: update $MY_OVERLAY"
  ```

See the [overlays guide](#overlays) to learn about the [overlays we provide](#provided-overlays) and [how to create your own overlays](#custom-overlays).

## Overlays

> ### About Overlays

Author: ```suggestion
```

> 

Author: ```suggestion
```

An [*overlay*](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/#bases-and-overlays) specifies customizations for a base directory of Kubernetes manifests, in this case the `base/` directory in the [reference repository](#reference-repository). 

Overlays can:
- Be used for example to change the number of replicas, change a namespace, add a label, etc
- Refer to other overlays that eventually refer to the base (forming a directed acyclic graph with the base as the root)

> Overlays can be used in one of two ways:

Author: ```suggestion
### Using overlays

Overlays can be used in one of two ways:
```

> 

Author: ```suggestion
```

command is a directory containing a `kustomization.yaml` file.
- With`kustomize`: This generates manifest files that are then applied in the conventional way using `kubectl apply -f`.

The overlays provided in our [overlays directory](https://github.com/sourcegraph/deploy-sourcegraph/tree/master/overlays) rely on the `kustomize` tool and the `overlay-generate-cluster.sh` script in the
`root` directory to generate the manifests. There are two reasons why it was set up like this:

- It avoids having to put a `kustomization.yaml` file in the `base` directory and forcing users that don't use overlays
  to deal with it (unfortunately `kubectl apply -f` doesn't work if a `kustomization.yaml` file is in the directory).
- It generates manifests instead of applying them directly. This provides opportunity to additionally validate the files
  and also allows using `kubectl apply -f` with `--prune` flag turned on (`apply -k` with `--prune` does not work correctly).

### Generating Manifests

To generate Kubernetes manifests from an overlay, run the `overlay-generate-cluster.sh` with two arguments:

> 

Author: ```suggestion
```

- the name of the overlay
- and a path to an output directory where the generated manifests will be

For example:

> 

Author: ```suggestion
```

```sh
#                overlay directory name    output directory
#                                 |             |
./overlay-generate-cluster.sh my-overlay generated-cluster
```

After executing the script you can apply the generated manifests from the `generated-cluster` directory:

```sh
kubectl apply --prune -l deploy=sourcegraph -f generated-cluster --recursive
```

We recommend that you:

> 

Author: ```suggestion
```

# Kustomize

## About Kustomize

> ### About Kustomize

Reviewer: ```suggestion
```

- It generates manifests instead of applying them directly. This provides opportunity to additionally validate the files
  and also allows using `kubectl apply -f` with `--prune` flag turned on (`apply -k` with `--prune` does not work correctly).

### Generating Manifests

To generate Kubernetes manifests from an overlay, run the `overlay-generate-cluster.sh` with two arguments:

- the name of the overlay
- and a path to an output directory where the generated manifests will be

For example:

```sh
#                overlay directory name    output directory
#                                 |             |
./overlay-generate-cluster.sh my-overlay generated-cluster
```

After executing the script you can apply the generated manifests from the `generated-cluster` directory:

```sh
kubectl apply --prune -l deploy=sourcegraph -f generated-cluster --recursive
```

We recommend that you:

- [Update the `./overlay-generate-cluster` script](./operations.md#applying-manifests) to apply the generated manifests from the `generated-cluster` directory with something like the above snippet

> - Commit your overlays changes separately - see our [customization guide](#customizations) for more details.

Author: I'm not sure what this link should go to. @caugustus-sourcegraph ?

This overlay adds a namespace declaration to all the manifests.

1. Change the namespace by replacing `ns-sourcegraph` to the name of your choice everywhere within the
[overlays/namespaced/](https://github.com/sourcegraph/deploy-sourcegraph/blob/master/overlays/namespaced/) directory. 

1. Generate the overlay by running this command from the `root` directory:

    ```
    ./overlay-generate-cluster.sh namespaced generated-cluster
    ```

1. Create the namespace if it doesn't exist yet:

    ```
    kubectl create namespace ns-<EXAMPLE NAMESPACE>
    kubectl label namespace ns-<EXAMPLE NAMESPACE> name=ns-sourcegraph
    ```

1. Apply the generated manifests (from the `generated-cluster` directory) by running this command from the `root` directory:

  ```
  kubectl apply -n ns-<EXAMPLE NAMESPACE> --prune -l deploy=sourcegraph -f generated-cluster --recursive
  ```

1. Check for the namespaces and their status with:

  ```
  kubectl get pods -A
  ```

#### Storageclass

By default Sourcegraph is configured to use a storage class called `sourcegraph`. If you wish to use an alternate name, you can use this overlay to change all `storageClass` references in the manifests. 

> You need to create the storageclass if it doesn't exist yet. See [these docs](#configure-a-storage-class) for more instructions.

Reviewer: ```suggestion
You need to create the storageclass if it doesn't exist yet. See [these docs](./configure.md#configure-a-storage-class) for more instructions.
```

./overlay-generate-cluster.sh migrate-to-nonprivileged generated-cluster
```

After executing the script you can apply the generated manifests from the generated-cluster directory:

```
kubectl apply --prune -l deploy=sourcegraph -f generated-cluster --recursive
```

#### minikube overlay

This kustomization deletes resource declarations and storage classnames to enable running Sourcegraph on minikube.

To use it, execute the following command from the `root` directory:

```sh
./overlay-generate-cluster.sh minikube generated-cluster
```

After executing the script you can apply the generated manifests from the generated-cluster directory:

```sh
minikube start
kubectl create namespace ns-sourcegraph
kubectl -n ns-sourcegraph apply --prune -l deploy=sourcegraph -f generated-cluster --recursive
kubectl -n ns-sourcegraph expose deployment sourcegraph-frontend --type=NodePort --name sourcegraph --port=3080 --target-port=3080
minikube service list
```

> NOTE: For Mac Users, run `minikube service sourcegraph -n ns-sourcegraph` to open the newly deployed Sourcegraph in your browser

To tear it down:

```sh
kubectl delete namespaces ns-sourcegraph
minikube stop
```

### Custom overlays

> To create your own [overlays](#overlays), first [set up your deployment reference repository to enable customizations](#getting-started).

Reviewer: ```suggestion
To create your own [overlays](#overlays), first [set up your deployment reference repository to enable customizations](./configure.md#getting-started).
```
